wurst/util/DialogBox.wurst

```diff
@@ -0,0 +1,84 @@
+package DialogBox
+import HashMap
+import Dialog
+import LinkedList
+
+public interface Callback
+    function run()
+
+/** DialogBox lets you to create dialogs and associate the buttons with code. */
+public class DialogBox
+
+    private static constant buttonClosures = new HashMap<button, Callback>
+    private static constant dialogBoxes = new HashMap<dialog, DialogBox>
+    private constant buttons = new LinkedList<button>
+
+    private constant dialogHandle = createDialog()
+    private constant clickDialog = CreateTrigger()
+
+    private string title    
+
+    construct(string title)
+        this.title = title
+        this.dialogHandle.setMessage(title)
+        dialogBoxes.put(this.dialogHandle, this)
+        this.clickDialog
+            ..registerDialogEvent(this.dialogHandle)
+            ..addAction(function removeClickedDialog)
+
+    private static function removeClickedDialog()
+        let btn = GetClickedButton()
+        let cb = buttonClosures.get(btn)
+        if(cb!=null)
+            cb.run()
+        destroy dialogBoxes.get(GetClickedDialog())
+
+    function addButton(string buttonText, Callback cb)
+        this.addButton(buttonText, 0 , cb)
+
+    /** Hotkey: use ASCII number of the capital letter. */
+    function addButton(string buttonText, integer hotkey, Callback cb)
+        let btn = this.dialogHandle.addButton(buttonText, hotkey)
+        this.buttons.add(btn)
+        buttonClosures.put(btn, cb)
```

Comment:
You can't add a null to a ```LinkedList<T>```, so this will throw an error. Check if ```cb``` is null. Unless, of course, this is intended behavior and adding null is an actual error (but I got a different idea from your argument that a user can willfully pass null). In that case, create additional ```addButton``` functions that take only text, and text and hotkey. Don't be afraid to overload functions many times if it makes the user not have to pass null to accomplish something. Even if the code looks a bit messier in the system, it will look neater in the user code, which is what you should aim for if you have 0-4 parameters like here.

Reply:
Thank you for your insight. I was considering overloading the function, but I was not sure if using null as parameter would be easier.
The problem here is, that there are 3 optional paramters, so this will be a lot of functions if every combination is a new function.
The linked list only contains the buttons, which are never null. Only the hashmap maps buttons to closures and the closure could be null. Can you put null values into hashmaps? At least there was not error message when I just tested it using a null closure.

Reply:
That's fine I guess, I didn't realize at first that it's a map.

Still, there are really not 3 optional parameters - there are only two. You will always need to name the button. This means that the optional parameters are going to be the hotkey and callback. This makes for a 2 * 2 = 4 functions which is okay to manage.

Reply:
Yes, I agree. I wasn't thinking about removing the string paramter, but somehow thought there were 3 other paramters. Probably because the quit button contains one additional paramter.